### PR TITLE
IGNITE-23290 Save and set a compaction revision into Metastorage

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-23290

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java [hunk 1]
```diff
@@ -835,6 +851,29 @@ public void advanceSafeTime(HybridTimestamp newSafeTime) {
         }
     }
 
+    @Override
+    public void saveCompactionRevision(long revision) {
+        throw new UnsupportedOperationException();
```
Comment (rpuch) [inline modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java:856]: Who will call this method?

File: modules/metastorage-api/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java [hunk 2]
```diff
@@ -296,9 +296,52 @@ public interface MetaStorageManager extends IgniteComponent {
      *     <li>Compaction revision is {@code 6}: "foo" [5].</li>
      * </ul>
      *
+     * <p>Compaction revision is expected to be less than the current metastorage revision.</p>
+     *
+     * <p>Compaction revision is not update or saved.</p>
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage-api/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java:301]: Thanks

File: modules/metastorage-api/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java [hunk 3]
```diff
@@ -296,9 +296,52 @@ public interface MetaStorageManager extends IgniteComponent {
...
      * @throws IgniteInternalException with cause {@link NodeStoppingException} if the node is in the process of stopping.
      * @throws MetaStorageException If there is an error during the metastorage compaction process.
      */
     void compactLocally(long revision);
+
+    /**
+     * Saves the compaction revision to the metastorage meta locally.
+     *
+     * <p>Method only saves the new compaction revision to the meta of metastorage. After invoking this method the metastorage read methods
+     * will <b>not</b> throw a {@link CompactedException} if they request a revision less than or equal to the new saved one.</p>
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage-api/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java:313]: thanks

File: modules/metastorage-api/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java [hunk 4]
```diff
@@ -296,9 +296,52 @@ public interface MetaStorageManager extends IgniteComponent {
...
+     *
+     * <p>Last saved compaction revision will be in the metastorage snapshot. When restore from a snapshot, compaction revision will be
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage-api/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java:315]: thanks

File: modules/metastorage-api/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java [hunk 5]
```diff
@@ -296,9 +296,52 @@ public interface MetaStorageManager extends IgniteComponent {
...
+     * restored after which the metastorage read methods will throw exception {@link CompactedException}.</p>
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage-api/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java:316]: thanks

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorage.java [hunk 6]
```diff
@@ -256,8 +257,12 @@ boolean invoke(
      *
      * <p>Compaction revision is expected to be less than the {@link #revision current storage revision}.</p>
      *
+     * <p>Compaction revision is not update or saved.</p>
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorage.java:260]: thanks

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorage.java [hunk 7]
```diff
@@ -324,4 +329,40 @@ boolean invoke(
...
      */
     void advanceSafeTime(HybridTimestamp newSafeTime);
+
+    /**
+     * Saves the compaction revision to the storage meta.
+     *
+     * <p>Method only saves the new compaction revision to the meta of storage. After invoking this method the metastorage read methods
+     * will <b>not</b> throw a {@link CompactedException} if they request a revision less than or equal to the new saved one.</p>
+     *
+     * <p>Last saved compaction revision will be in the {@link #snapshot snapshot}. When {@link #restoreSnapshot restore} from a snapshot,
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorage.java:339]: thnaks

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java [hunk 8]
```diff
@@ -64,4 +64,12 @@ public static int indexToCompact(long[] keyRevisions, long compactionRevisionInc
     public static String toUtf8String(byte[] bytes) {
         return new String(bytes, UTF_8);
     }
+
+    /** Asserts that the compaction revision is less than the current repository revision. */
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java:68]: Fix it

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java [hunk 9]
```diff
@@ -64,4 +64,12 @@ public static int indexToCompact(long[] keyRevisions, long compactionRevisionInc
...
+    public static void assertCompactionRevisionLessCurrent(long compactionRevision, long revision) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/KeyValueStorageUtils.java:69]: Thanks

File: modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java [hunk 10]
```diff
@@ -143,6 +160,104 @@ void testCompactRevisionSequentially() {
...
+    @ValueSource(booleans = {true, false})
+    void testSetAndGetCompactionRevisionAndRestart(boolean clearStorage) throws Exception {
+        storage.setCompactionRevision(1);
+
+        restartStorage(clearStorage);
+        assertEquals(-1, storage.getCompactionRevision());
+    }
+
+    @Test
+    void testSaveCompactionRevision() {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage/src/test/java/org/apache/ignite/internal/metastorage/server/AbstractCompactionKeyValueStorageTest.java:184]: thanks

File: modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java (hunk 1 above)
Comment (tkalkirill) [reply modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java:856]: Nobody I suppose, just like the snapshots in the storage.

File: modules/metastorage-api/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java [hunk 11]
```diff
@@ -296,9 +296,52 @@ public interface MetaStorageManager extends IgniteComponent {
...
+     * Sets the compaction revision locally, but does not save it, after invoking this method the metastorage read methods will throw a
+     * {@link CompactedException} if they request a revision less than or equal to the new one.
+     *
+     * <p>Compaction revision is expected to be less than the current metastorage revision.</p>
+     *
+     * @param revision Compaction revision.
+     * @throws IgniteInternalException with cause {@link NodeStoppingException} if the node is in the process of stopping.
+     * @see #saveCompactionRevisionLocally(long)
+     */
+    void setCompactionRevisionLocally(long revision);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/metastorage-api/src/main/java/org/apache/ignite/internal/metastorage/MetaStorageManager.java:337]: I would leave it as is for now, we will rename it later if necessary.

File: modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java (hunk 1 above)
Comment (rpuch) [reply modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java:858]: If it's only needed in the RocksDB-related implementation, I suggest calling it directly and not via interface method (and remove the method from the interface)

File: modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java [hunk 12]
```diff
@@ -1677,9 +1673,4 @@ private boolean isTombstoneForCompaction(byte[] key, long revision) {
             );
         }
     }
-
-    private static WriteOptions createWriteOptions() {
-        // Metastorage recovery is based on the snapshot & external log. WAL is never used for recovery, and can be safely disabled.
-        return new WriteOptions().setDisableWAL(true);
```
Comment (rpuch) [inline modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java]: Why isn't WAL disabled anymore?

Comment (tkalkirill) [reply modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java]: Everything is okay now it’s in `org.apache.ignite.internal.metastorage.server.persistence.RocksDbKeyValueStorage#defaultWriteOptions`, it came from another PR.

```
    /** Metastorage recovery is based on the snapshot & external log. WAL is never used for recovery, and can be safely disabled. */
    private final WriteOptions defaultWriteOptions = new WriteOptions().setDisableWAL(true);
```

File: modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java (hunk 1 above)
Comment (tkalkirill) [reply modules/metastorage/src/testFixtures/java/org/apache/ignite/internal/metastorage/server/SimpleInMemoryKeyValueStorage.java:872]: It seems right to me that this should be in the interface. For example, there are methods `org.apache.ignite.internal.metastorage.server.SimpleInMemoryKeyValueStorage#snapshot` and `org.apache.ignite.internal.metastorage.server.SimpleInMemoryKeyValueStorage#snapshot` that also do not work for the test storage. If anything, we will fix it later.

**Review by rpuch — APPROVED** [review]

**Pull request merged**